src/bio/bio_config.c

```diff
@@ -0,0 +1,531 @@
+			     SPDK_JSON_PARSE_FLAG_ALLOW_COMMENTS);
+	if (rc < 0) {
+		D_ERROR("Parsing JSON configuration failed (%zd)\n", rc);
+		rc = -DER_INVAL;
+		goto err;
+	}
+
+	values_cnt = rc;
+	values = calloc(values_cnt, sizeof(struct spdk_json_val));
+	if (values == NULL) {
+		D_ERROR("Out of memory\n");
+		rc = -DER_NOMEM;
+		goto err;
+	}
+
+	rc = spdk_json_parse(json, json_size, values, values_cnt, &end,
+			     SPDK_JSON_PARSE_FLAG_ALLOW_COMMENTS);
+	if (rc != values_cnt) {
+		D_ERROR("Parsing JSON configuration failed (%zd)\n", rc);
+		rc = -DER_INVAL;
+		goto err;
+	}
+
+	ctx->json_data = json;
+	ctx->json_data_size = json_size;
+
+	ctx->values = values;
+	ctx->values_cnt = values_cnt;
+
+	return 0;
+err:
+	D_FREE(json);
+	D_FREE(values);
+	return rc;
+}
+
+static int
+load_vmd_subsystem_config(struct json_config_ctx *ctx, bool *vmd_enabled)
+{
+	struct config_entry	 cfg = {};
+
+	D_ASSERT(ctx->config_it != NULL);
+	D_ASSERT(vmd_enabled != NULL);
+
+	if (spdk_json_decode_object(ctx->config_it, jsonrpc_cmd_decoders,
+				    SPDK_COUNTOF(jsonrpc_cmd_decoders), &cfg)) {
+		D_ERROR("Failed to decode config entry\n");
+		return -DER_INVAL;
+	}
+
+	if (strcmp(cfg.method, "enable_vmd") == 0)
+		*vmd_enabled = true;
+
+	D_FREE(cfg.method);
+	return 0;
+}
+
+static int
+load_bdev_subsystem_config(struct json_config_ctx *ctx, bool vmd_enabled,
+			   struct spdk_env_opts *opts)
+{
+	struct config_entry	 cfg = {};
+	struct spdk_json_val	*key;
+	char			*traddr;
+	int			 rc = 0;
+
+	D_ASSERT(ctx->config_it != NULL);
+
+	if (spdk_json_decode_object(ctx->config_it, jsonrpc_cmd_decoders,
+				    SPDK_COUNTOF(jsonrpc_cmd_decoders), &cfg)) {
+		D_ERROR("Failed to decode config entry\n");
+		return -DER_INVAL;
+	}
+
+	if ((strcmp(cfg.method, "bdev_nvme_attach_controller") != 0) || (cfg.params == NULL))
+		goto out;
+
+	key = spdk_json_object_first(cfg.params);
+	while (key != NULL) {
+		if (spdk_json_strequal(key, "traddr")) {
+			traddr = spdk_json_strdup(json_value(key));
+			D_INFO("Transport address found in JSON config: %s\n", traddr);
+
+			if (vmd_enabled) {
+				if (strncmp(traddr, "0", 1) != 0) {
+					/*
+					 * We can assume this is the transport id of the
+					 * backing NVMe SSD behind the VMD. DPDK will
+					 * not recognize this transport ID, instead need
+					 * to pass VMD address as the whitelist param.
+					 */
+					rc = traddr_to_vmd(traddr, traddr);
+					if (rc != 0) {
+						D_ERROR("Invalid traddr=%s\n", traddr);
+						rc = -DER_INVAL;
+						goto out;
+					}
+					D_INFO("\tVMD backing address translated to: %s\n", traddr);
+				}
+			}
+
+			rc = opts_add_pci_addr(opts, traddr);
+			if (rc != 0) {
+				D_ERROR("spdk env add pci: %d\n", rc);
+				goto out;
+			}
+		}
+
+		key = spdk_json_next(key);
+	}
+out:
+	D_FREE(cfg.method);
+	return rc;
+}
+
+static int
+add_bdevs_to_opts(struct spdk_json_val *bdev_ss, bool vmd_enabled, struct spdk_env_opts *opts)
+{
+	struct json_config_ctx	*ctx = calloc(1, sizeof(*ctx));
```

Comment:
I don't see why a new 'ctx' is required here, is it because you didn't do the subsystem decoding in the subsystems loop? Probably you'd do the decoding in the loop where the cursors in 'json_config_ctx' are correct.